src/main/scala/za/co/absa/hyperdrive/trigger/persistance/WorkflowRepository.scala

```diff
@@ -19,43 +19,55 @@ import java.time.LocalDateTime
   def getWorkflowsByProjectName(projectName: String)(implicit ec: ExecutionContext): Future[Seq[Workflow]]
-  def deleteWorkflow(id: Long)(implicit ec: ExecutionContext): Future[Unit]
-  def updateWorkflow(workflow: WorkflowJoined)(implicit ec: ExecutionContext): Future[Either[ApiError, Unit]]
-  def switchWorkflowActiveState(id: Long)(implicit ec: ExecutionContext): Future[Unit]
+  def deleteWorkflow(id: Long, user: String)(implicit ec: ExecutionContext): Future[Unit]
+  def updateWorkflow(workflow: WorkflowJoined, user: String)(implicit ec: ExecutionContext): Future[Either[ApiError, Unit]]
+  def switchWorkflowActiveState(id: Long, user: String)(implicit ec: ExecutionContext): Future[Unit]
   def getProjects()(implicit ec: ExecutionContext): Future[Seq[String]]
   def getProjectsInfo()(implicit ec: ExecutionContext): Future[Seq[ProjectInfo]]
 }
 
 @stereotype.Repository
-class WorkflowRepositoryImpl extends WorkflowRepository {
+class WorkflowRepositoryImpl(override val workflowHistoryRepository: WorkflowHistoryRepository) extends WorkflowRepository {
+
   import profile.api._
+
   private val logger = LoggerFactory.getLogger(this.getClass)
 
-  override def insertWorkflow(workflow: WorkflowJoined)(implicit ec: ExecutionContext): Future[Either[ApiError, Long]] = db.run(
-    (for {
-      workflowId <- workflowTable returning workflowTable.map(_.id) += workflow.toWorkflow.copy(created = LocalDateTime.now())
-      sensorId <- sensorTable += workflow.sensor.copy(workflowId = workflowId)
-      dagId <- dagDefinitionTable returning dagDefinitionTable.map(_.id) += workflow.dagDefinitionJoined.toDag().copy(workflowId = workflowId)
-      jobId <- jobDefinitionTable ++= workflow.dagDefinitionJoined.jobDefinitions.map(_.copy(dagDefinitionId = dagId))
-    } yield (workflowId)).transactionally.asTry
-  ).map {
-    case Success(workflowId) => Right(workflowId)
-    case Failure(ex) =>
-      logger.error(s"Unexpected error occurred when inserting workflow $workflow", ex)
-      Left(GenericDatabaseError)
+  override def insertWorkflow(workflow: WorkflowJoined, user: String)(implicit ec: ExecutionContext): Future[Either[ApiError, Long]] = {
+    db.run(
+      (for {
+        workflowId <- workflowTable returning workflowTable.map(_.id) += workflow.toWorkflow.copy(created = LocalDateTime.now())
+        sensorId <- sensorTable returning sensorTable.map(_.id) += workflow.sensor.copy(workflowId = workflowId)
+        dagId <- dagDefinitionTable returning dagDefinitionTable.map(_.id) += workflow.dagDefinitionJoined.toDag().copy(workflowId = workflowId)
+        jobId <- jobDefinitionTable returning jobDefinitionTable.map(_.id) ++= workflow.dagDefinitionJoined.jobDefinitions.map(_.copy(dagDefinitionId = dagId))
+      } yield {
+        workflowId
+      }).flatMap(id => {
+        getWorkflowJoined(id).map(
+          workflowUpdated => workflowHistoryRepository.create(workflowUpdated, user)
```

Comment:
Yes, I can but I wanted to have it separated from insert functionality ( and potentialy we can extract it into function and call only that one). Yes, we need exact copy which is updated/created (on input you already have out of date workflow, because  of .copy(created = LocalDateTime.now()). Even if I update workflow manually I will always have workflow which in db).